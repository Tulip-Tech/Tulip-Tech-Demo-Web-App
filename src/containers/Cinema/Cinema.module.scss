@use 'src/styles/variables';
@use 'src/styles/theme';
@use 'src/styles/mixins/responsive';
@use 'src/styles/mixins/typography';

.fade {
  position: relative;
  // make sure the fade animation happens on top of page elements
  z-index: variables.$player-z-index;
}

.cinema {
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: calc(100vh - calc(100vh - 100%));
  overflow: hidden;
  background-color: variables.$black;
}

.playerOverlay {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  min-height: 100px;
  padding: 24px 56px;
  background-image: linear-gradient(to top, rgba(0, 0, 0, 0) 0%, rgba(0, 0, 0, 0.1) 25%, rgba(0, 0, 0, 0.25) 100%);
  pointer-events: none;

  @include responsive.mobile-only() {
    padding-left: 16px;
  }
}

.playerContent {
  display: flex;
  width: 80%;

  @include responsive.mobile-only() {
    width: 100%;
  }
}

.backButton {
  margin-right: 24px;
  opacity: 1;
  pointer-events: auto;

  > svg {
    width: 36px;
    height: 36px;
    -webkit-filter: drop-shadow(2px 2px 5px rgba(0, 0, 0, 0.4)); /* stylelint-disable-line */
    filter: drop-shadow(2px 2px 5px rgba(0, 0, 0, 0.4));
  }
}

.title {
  @include typography.video-title-base;
}

.metaContainer {
  display: flex;
  flex-direction: column;
}

.primaryMetadata {
  margin-bottom: 8px;
  font-size: 16px;
  line-height: 18px;
  letter-spacing: 0.15px;

  @include responsive.mobile-only() {
    order: 2;
    font-size: 14px;
  }
}

.secondaryMetadata {
  margin-top: 4px;
  margin-bottom: 8px;
  font-size: 20px;
  line-height: 20px;
  letter-spacing: 0.5px;

  @include responsive.mobile-only() {
    order: 1;
    margin: 4px 0;
    font-size: 18px;
  }
}
