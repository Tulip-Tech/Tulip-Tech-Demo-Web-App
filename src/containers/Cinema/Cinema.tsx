import React, { useCallback, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';

import styles from './Cinema.module.scss';

import type { PlaylistItem } from '#types/playlist';
import Fade from '#components/Animation/Fade/Fade';
import IconButton from '#components/IconButton/IconButton';
import ArrowLeft from '#src/icons/ArrowLeft';
import PlayerContainer from '#src/containers/PlayerContainer/PlayerContainer';

type Props = {
  open: boolean;
  item: PlaylistItem;
  onPlay?: () => void;
  onPause?: () => void;
  onComplete?: () => void;
  onClose?: () => void;
  onNext?: () => void;
  feedId?: string;
  title: string;
  primaryMetadata: React.ReactNode;
  secondaryMetadata?: React.ReactNode;
  liveStartDateTime?: string | null;
  liveEndDateTime?: string | null;
  liveFromBeginning?: boolean;
};

const Cinema: React.FC<Props> = ({
  open,
  item,
  title,
  primaryMetadata,
  secondaryMetadata,
  onPlay,
  onPause,
  onComplete,
  onClose,
  onNext,
  feedId,
  liveStartDateTime,
  liveEndDateTime,
  liveFromBeginning,
}: Props) => {
  const { t } = useTranslation();

  // state
  const [isPlaying, setIsPlaying] = useState<boolean>(false);
  const [userActive, setUserActive] = useState(true);

  const handlePlay = useCallback(() => {
    setIsPlaying(true);
    onPlay && onPlay();
  }, [onPlay]);

  const handlePause = useCallback(() => {
    setIsPlaying(false);
    onPause && onPause();
  }, [onPause]);

  const handleComplete = useCallback(() => {
    onComplete && onComplete();
  }, [onComplete]);

  const handleNext = useCallback(() => {
    onNext && onNext();
  }, [onNext]);

  const handleUserActive = useCallback(() => setUserActive(true), []);
  const handleUserInactive = useCallback(() => setUserActive(false), []);

  // effects
  useEffect(() => {
    if (open) {
      setUserActive(true);
      document.body.style.overflowY = 'hidden';
    }

    return () => {
      document.body.style.overflowY = '';
    };
  }, [open]);

  return (
    <Fade open={open} className={styles.fade}>
      <div className={styles.cinema}>
        <PlayerContainer
          item={item}
          feedId={feedId}
          autostart={true}
          onPlay={handlePlay}
          onPause={handlePause}
          onComplete={handleComplete}
          onUserActive={handleUserActive}
          onUserInActive={handleUserInactive}
          onNext={handleNext}
          liveEndDateTime={liveEndDateTime}
          liveFromBeginning={liveFromBeginning}
          liveStartDateTime={liveStartDateTime}
        />

        <Fade open={!isPlaying || userActive}>
          <div className={styles.playerOverlay}>
            <div className={styles.playerContent}>
              <IconButton aria-label={t('common:back')} onClick={onClose} className={styles.backButton}>
                <ArrowLeft />
              </IconButton>
              <div>
                <h2 className={styles.title}>{title}</h2>
                <div className={styles.metaContainer}>
                  {secondaryMetadata && <div className={styles.secondaryMetadata}>{secondaryMetadata}</div>}
                  <div className={styles.primaryMetadata}>{primaryMetadata}</div>
                </div>
              </div>
            </div>
          </div>
        </Fade>
      </div>
    </Fade>
  );
};

export default Cinema;
