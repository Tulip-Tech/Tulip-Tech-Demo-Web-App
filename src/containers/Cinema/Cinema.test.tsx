import React from 'react';

import Cinema from './Cinema';

import { renderWithRouter } from '#test/testUtils';
import type { PlaylistItem } from '#types/playlist';

describe('<Cinema>', () => {
  test('renders and matches snapshot', () => {
    const item = {
      description: 'Test item description',
      duration: 354,
      feedid: 'ax85aa',
      image: 'http://test/img.jpg',
      images: [],
      link: 'http://test/link',
      genre: 'Tester',
      mediaid: 'zp50pz',
      pubdate: 26092021,
      rating: 'CC_CC',
      sources: [],
      seriesId: 'ag94ag',
      tags: 'Test tag',
      title: 'Test item title',
      tracks: [],
    } as PlaylistItem;
    const { container } = renderWithRouter(
      <Cinema item={item} onPlay={() => null} onPause={() => null} open={true} title={item.title} primaryMetadata="Primary metadata" />,
    );

    expect(container).toMatchSnapshot();
  });
});
