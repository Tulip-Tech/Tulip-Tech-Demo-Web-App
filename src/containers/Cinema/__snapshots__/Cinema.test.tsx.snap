// Vitest Snapshot v1

exports[`<Cinema> > renders and matches snapshot 1`] = `
<div>
  <div
    class="_fade_555f3c"
    style="transition: opacity 0.25s ease-in-out; opacity: 0;"
  >
    <div
      class="_cinema_555f3c"
    >
      <div
        class="_container_20cd38"
        data-testid="player-container"
      >
        <div />
      </div>
      <div
        style="transition: opacity 0.25s ease-in-out; opacity: 0;"
      >
        <div
          class="_playerOverlay_555f3c"
        >
          <div
            class="_playerContent_555f3c"
          >
            <div
              aria-label="common:back"
              class="_iconButton_0fef65 _backButton_555f3c"
              role="button"
              tabindex="0"
            >
              <svg
                aria-hidden="true"
                class="_icon_1e9999"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
                  />
                </g>
              </svg>
            </div>
            <div>
              <h2
                class="_title_555f3c"
              >
                Test item title
              </h2>
              <div
                class="_metaContainer_555f3c"
              >
                <div
                  class="_primaryMetadata_555f3c"
                >
                  Primary metadata
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
